srw/direct_access.cpp

```diff
@@ -0,0 +1,205 @@
+/*
+ * 2014+ Copyright (c) Yandex
+ * All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ */
+
+#include <memory>
+#include <iostream>
+#include <fstream>
+
+#include <cocaine/context.hpp>
+#include <cocaine/logging.hpp>
+
+#include <elliptics/interface.h>
+#include <elliptics/srw.h>
+#include <elliptics/utils.hpp>
+
+#include <library/elliptics.h>
+
+#include "direct_access.hpp"
+
+using namespace cocaine;
+using namespace cocaine::io;
+using namespace ioremap::elliptics;
+
+using namespace std::placeholders;
+
+namespace {
+
+std::string to_string(const std::vector<int> &v)
+{
+    std::stringstream ss;
+    for (size_t i = 0; i < v.size(); ++i) {
+        if (i > 0) {
+            ss << ", ";
+        }
+        ss << v[i];
+    }
+    return ss.str();
+}
+
+}
+
+namespace ioremap { namespace elliptics {
+
+// Is there a better way to find groups that are served by this node?
+std::vector<int> find_local_groups(session &sess)
+{
+	const auto &routes = sess.get_routes();
+	const dnet_node *node = sess.get_native_node();
+	std::set<int> unique_groups;
+	for (const auto &route : routes) {
+		for (int i = 0; i < node->addr_num; ++i) {
+			if (dnet_addr_equal(&route.addr, &node->addrs[i])) {
+				unique_groups.insert(route.group_id);
+			}
+		}
+	}
+	return std::move(std::vector<int>(unique_groups.begin(), unique_groups.end()));
+}
+
+direct_access_t::direct_access_t(cocaine::context_t& context, cocaine::io::reactor_t& reactor, const std::string& name, const Json::Value& args, dnet_node* node)
+	: service_t(context, reactor, name, args)
+	, m_log(context, name)
+	, m_session_proto(node)
+{
+	COCAINE_LOG_INFO((&m_log), "%s: enter", __func__);
+
+	on<direct_access::read_data>("read_data", std::bind(&direct_access_t::read_data, this, _1, _2, _3));
+	on<direct_access::lookup>("lookup", std::bind(&direct_access_t::lookup, this, _1));
+	on<direct_access::write_data>("write_data", std::bind(&direct_access_t::write_data, this, _1, _2, _3));
+
+	// find all groups that resides on our node, and set them as session default
+	m_session_proto.set_groups(find_local_groups(m_session_proto));
```

Comment:
This will force `m_session_proto` to perform action on every single group which is present in current server's config.
It is probably not intended behaviour